bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/HandleFactoryImpl.java

```diff
@@ -21,14 +21,25 @@
 
 package org.apache.bookkeeper.bookie;
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
 import java.io.IOException;
+import java.time.Duration;
 import org.apache.bookkeeper.bookie.LedgerStorage.LedgerDeletionListener;
 import org.apache.bookkeeper.util.collections.ConcurrentLongHashMap;
 
 class HandleFactoryImpl implements HandleFactory, LedgerDeletionListener {
     private final ConcurrentLongHashMap<LedgerDescriptor> ledgers;
     private final ConcurrentLongHashMap<LedgerDescriptor> readOnlyLedgers;
 
+    /**
+     * Once the ledger was marked "fenced" before, the ledger was accessed by multi clients. One client is calling
+     * "delete" now, and other clients may call "write" continuously later. We mark these ledgers can not be written
+     * anymore. And maintains the state for 7 days is safety.
+     */
+    private final Cache<Long, Boolean> recentlyFencedAndDeletedLedgers = CacheBuilder.newBuilder()
```

Comment:
This is in memory, what happens if you restart the bookie ?

Reply:
> This is in memory, what happens if you restart the bookie?

I think scenarios `1` and `2` below are fine, and scenario `3` below is not fine, do you have any suggestions?

---

**Scenario 1: the journal log is still there**
- Deleted the ledger.
- Restart BK.
- Replay journal logs.
- Reset the ledger as fenced.
  - Since the ledger is fenced, the client will get a fenced error when it tries to write. 
- Delete the ledger after a GC
  - Move the ledger into  `recentlyFencedAndDeletedLedgers`
  - Since the ledger is in the variable `recentlyFencedAndDeletedLedgers`, the client will get a fenced error when it tries to write.

---

**Scenario 2: the journal log was removed & the client switched the ensemble due to pending operations failed**
- Deleted the ledger.
  - Journal logs were removed.
- Restart BK.
- The client tries to switch the ensemble due to previous operation fails
  - Since the ledger metadata was removed, this operation will fail.

---

 **Scenario 3: the journal log was removed & the client did not switch the ensemble because there are no pending operations**
- Deleted the ledger.
  - Journal logs were removed.
- Restart BK.
- The client reconnect successfully.
- If the client tries to write, it will be successful.

Reply:
@eolivelli How about we split this into 2 PRs
- Fix it in memory, which is what the current PR did. And this PR can be cherry-picked into the previous branches, this fix can fix almost all the scenarios(the probably of brokers and Bks crashes at the same time is quite few)
- Fix it with persistent storage, we can persist the states into a file, and add a config to define which file will be used, this will be released with the next feature release